In [1]:
print('hello world ')

hello world 


- pandas = adds data structure and toold designed to work with table like dtara ( similar to series )
- allow to handel missing data 
- provide tools like reshaping , merging sorting slicing and indexing 

- numpy =  should work on arithmatic operation& array operation
- pandas = data manipulation = data frame = 3d  and series = 2d 
- scikit = used for mechie learning operaton
- keras = used for deep learing operation/ tensor flow 

VIZUALTION LIBARIES 

- matpolotlib = is for basic vizualization
- seaborn = for advance vizulation 
- ploty = is very rarely use depends on numpy and pandas 


In [2]:
pip install pandas


Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [3]:
import pandas as pd 
import numpy as np 


In [4]:
pip install matplotlib

Defaulting to user installation because normal site-packages is not writeable

[notice] A new release of pip is available: 24.2 -> 24.3.1
[notice] To update, run: pip3 install --upgrade pip
Note: you may need to restart the kernel to use updated packages.


In [5]:
import matplotlib.pyplot as plt

In [6]:
df=pd.read_csv('Salaries.csv')
df.head(10)

rank discipline  phd  service   sex  salary
0       Prof          B   56       49  Male  186960
1       Prof          A   12        6  Male   93000
2       Prof          A   23       20  Male  110515
3       Prof          A   40       31  Male  131205
4       Prof          B   20       18  Male  104800
5       Prof          A   20       20  Male  122400
6  AssocProf          A   20       17  Male   81285
7       Prof          A   18       18  Male  126300
8       Prof          A   29       19  Male   94350
9       Prof          A   51       51  Male   57800

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   rank        78 non-null     object
 1   discipline  78 non-null     object
 2   phd         78 non-null     int64 
 3   service     78 non-null     int64 
 4   sex         78 non-null     object
 5   salary      78 non-null     int64 
dtypes: int64(3), object(3)
memory usage: 3.8+ KB


In [8]:
help(pd.read_csv)

Help on function read_csv in module pandas.io.parsers.readers:

read_csv(filepath_or_buffer: 'FilePath | ReadCsvBuffer[bytes] | ReadCsvBuffer[str]', *, sep: 'str | None | lib.NoDefault' = <no_default>, delimiter: 'str | None | lib.NoDefault' = None, header: "int | Sequence[int] | None | Literal['infer']" = 'infer', names: 'Sequence[Hashable] | None | lib.NoDefault' = <no_default>, index_col: 'IndexLabel | Literal[False] | None' = None, usecols: 'UsecolsArgType' = None, dtype: 'DtypeArg | None' = None, engine: 'CSVEngine | None' = None, converters: 'Mapping[Hashable, Callable] | None' = None, true_values: 'list | None' = None, false_values: 'list | None' = None, skipinitialspace: 'bool' = False, skiprows: 'list[int] | int | Callable[[Hashable], bool] | None' = None, skipfooter: 'int' = 0, nrows: 'int | None' = None, na_values: 'Hashable | Iterable[Hashable] | Mapping[Hashable, Iterable[Hashable]] | None' = None, keep_default_na: 'bool' = True, na_filter: 'bool' = True, verbose: 'bool | 

In [9]:
df.head(10)# frist 10 observaitons
df.tail(10)# lats 10 observations
print(df.shape) # tell us no of rows and columns 

(78, 6)


In [10]:
df.columns # tell us no of coloums 

Index(['rank', 'discipline', 'phd', 'service', 'sex', 'salary'], dtype='object')

In [11]:
df.dtypes # tell us data types of each coloum 

rank          object
discipline    object
phd            int64
service        int64
sex           object
salary         int64
dtype: object

- tell us data types of each coloum 
- from above object  means  any categorical data refer as object or others are number so int64 or float64 

In [12]:
# to accuss to a perticular coloum 
df['salary'] # here u can use both single court or double count 

0     186960
1      93000
2     110515
3     131205
4     104800
       ...  
73    105450
74    104542
75    124312
76    109954
77    109646
Name: salary, Length: 78, dtype: int64

In [13]:
df["rank"].dtype # to access to ranks coloum

dtype('O')

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 78 entries, 0 to 77
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   rank        78 non-null     object
 1   discipline  78 non-null     object
 2   phd         78 non-null     int64 
 3   service     78 non-null     int64 
 4   sex         78 non-null     object
 5   salary      78 non-null     int64 
dtypes: int64(3), object(3)
memory usage: 3.8+ KB


- suppose in rsnt we get 75 non null means 2 null valaues are there from above 

In [15]:
df.axes

[RangeIndex(start=0, stop=78, step=1),
 Index(['rank', 'discipline', 'phd', 'service', 'sex', 'salary'], dtype='object')]

In [16]:
df.values

array([['Prof', 'B', 56, 49, 'Male', 186960],
       ['Prof', 'A', 12, 6, 'Male', 93000],
       ['Prof', 'A', 23, 20, 'Male', 110515],
       ['Prof', 'A', 40, 31, 'Male', 131205],
       ['Prof', 'B', 20, 18, 'Male', 104800],
       ['Prof', 'A', 20, 20, 'Male', 122400],
       ['AssocProf', 'A', 20, 17, 'Male', 81285],
       ['Prof', 'A', 18, 18, 'Male', 126300],
       ['Prof', 'A', 29, 19, 'Male', 94350],
       ['Prof', 'A', 51, 51, 'Male', 57800],
       ['Prof', 'B', 39, 33, 'Male', 128250],
       ['Prof', 'B', 23, 23, 'Male', 134778],
       ['AsstProf', 'B', 1, 0, 'Male', 88000],
       ['Prof', 'B', 35, 33, 'Male', 162200],
       ['Prof', 'B', 25, 19, 'Male', 153750],
       ['Prof', 'B', 17, 3, 'Male', 150480],
       ['AsstProf', 'B', 8, 3, 'Male', 75044],
       ['AsstProf', 'B', 4, 0, 'Male', 92000],
       ['Prof', 'A', 19, 7, 'Male', 107300],
       ['Prof', 'A', 29, 27, 'Male', 150500],
       ['AsstProf', 'B', 4, 4, 'Male', 92000],
       ['Prof', 'A', 33, 30, 'Ma

 DATA TYPES


- 1 OBJECT = STRING = categorical data
- 2 int64 = integer data = numbers where 64 is the number allocated to the memory 
- 3 flot = decimal data = numbers where it have decimal values like 2.4,3.5 etc 
- 4 datatime64  = data and time in python standard library = look into these time series and expermemnt 


 DATA FRAME AND ATTRIBUTES 


- dtype   = list the type of coloumns
- coloums =list the coloumns names 
- axes    = list the row lables and coloumns names 
- ndim    = number of demensions 
- size    = number of elementss
- shape   = number of rows and coloumns 
- values  = numpy representation of the data 

- if we look in shape the out put is (78,6) if u see the data in mormal parentesis is called tuple if u see the data in square brackets is called list
- wher tuple are immutable and list are mutable ,mutable means we can change the data in the list but in tuple we cannot change the data 

EXAMPLE 

- find how many records this data frames has        = df.shape[]
- how many coloumns are there                       = df.shape[]
- what are the colooumns names                     = df.coloumns
- what type of coloumns we have in thid data frame =df.dtypes 

DATA FRAME METHODS 


- head(n),tail(n)     = top no of rows ,last no of rows 
- descride()          = it generate descriptive statistics for only numerical coloums  here descriptive statistics means mean median and mode ,standaead deviation etc 
- max(),min()         = it gives the max and min values of each coloumns for numeriacial data 
- mean(), median()   = to find the mean and median of each coloumns 
- std()               = to find the standard deviation of each coloumns 
- sample(n)           = to find the sample of n rows 
- dropna()            = to drop the missing values 

In [18]:
df.describe()

phd    service         salary
count  78.000000  78.000000      78.000000
mean   19.705128  15.051282  108023.782051
std    12.498425  12.139768   28293.661022
min     1.000000   0.000000   57800.000000
25%    10.250000   5.250000   88612.500000
50%    18.500000  14.500000  104671.000000
75%    27.750000  20.750000  126774.750000
max    56.000000  51.000000  186960.000000

- here 50% is the median value
- min and max without removing the outliers



In [19]:
# Check the data types of the DataFrame columns
print(df.dtypes)

rank          object
discipline    object
phd            int64
service        int64
sex           object
salary         int64
dtype: object


In [20]:
# Select only numeric columns
numeric_df=df.select_dtypes(include=[np.number])
# Calculate skewness for numeric columns
skewness=numeric_df.skew()
print(skewness)

phd        0.634366
service    0.913750
salary     0.452103
dtype: float64


 df.skew() is a method used to compute the skewness of the distribution of each column in a DataFrame

 A skewness value of 0 indicates a perfectly symmetrical distribution

General Rules for Interpreting Skewness:

1. Skewness ≈ 0: The distribution is nearly symmetric.

2.Positive Skewness (> 0): The distribution has a long tail on the right (most values are concentrated on the left).

3.Negative Skewness (< 0): The distribution has a long tail on the left (most values are concentrated on the right).

4.Magnitude:
- Small Skewness (< 0.5): Distribution is approximately symmetric.
- Moderate Skewness (0.5 to 1): Distribution is moderately skewed.
- High Skewness (> 1): Distribution is highly skewed.


 --------------------- OR --------------------




 Column A (0.634366): This value is positive but close to 0. This means the distribution of the values in Column A is slightly right-skewed, but not extremely so. There is a mild tail on the right side, but it’s not very pronounced.

Column B (0.913750): This value is positive and a bit larger than Column A. It suggests that the values in Column B have a moderate right skew—there is still a tail to the right, but not a very extreme one.

Column C (0.452103): This value is positive but closer to 0 again. Similar to Column A, it suggests that Column C is slightly right-skewed.

How to interpret positive skewness values:
Near 0 (like 0.634366 or 0.452103): The distribution is nearly symmetric, but with a slight tail on the right side.
Above 1: A stronger right-skew, meaning the right tail is more noticeable.
Negative values: Would indicate left-skewness (tail on the left side).

In [21]:
print(df.dtypes)

rank          object
discipline    object
phd            int64
service        int64
sex           object
salary         int64
dtype: object


In [22]:
kurtosis= numeric_df.kurt()
print(kurtosis)

phd        0.042504
service    0.608981
salary    -0.401713
dtype: float64


here what we did is for we already seperated numerical coloumns into numeric_df in previous skew now we use the same numeric_df for kurt so """kurtosis= numeric_df.kurt()""""

In [23]:
numeric_df.mean()

phd            19.705128
service        15.051282
salary     108023.782051
dtype: float64

In [25]:
df[["phd","service"]].describe() # we used double bracket because we are using two coloumns 


phd    service
count  78.000000  78.000000
mean   19.705128  15.051282
std    12.498425  12.139768
min     1.000000   0.000000
25%    10.250000   5.250000
50%    18.500000  14.500000
75%    27.750000  20.750000
max    56.000000  51.000000

In [26]:
df['sex'].value_counts()

sex
Male      39
Female    39
Name: count, dtype: int64

In [27]:
df['sex'].unique()

array(['Male', 'Female'], dtype=object)

In [28]:
df['phd'] 

0     56
1     12
2     23
3     40
4     20
      ..
73    18
74    19
75    17
76    28
77    23
Name: phd, Length: 78, dtype: int64

In [31]:
df['rank'].value_counts()

rank
Prof         46
AsstProf     19
AssocProf    13
Name: count, dtype: int64

In [33]:
df.max()

rank            Prof
discipline         B
phd               56
service           51
sex             Male
salary        186960
dtype: object

In [34]:
df.min()

rank          AssocProf
discipline            A
phd                   1
service               0
sex              Female
salary            57800
dtype: object

In [39]:
df_50=df.sample(50) # here we stored 50 random samples in df_50
df_50# print the 50 saples 


rank discipline  phd  service     sex  salary
40       Prof          A   39       36  Female  137000
43   AsstProf          B    5        0  Female   77000
16   AsstProf          B    8        3    Male   75044
26       Prof          A   38       19    Male  148750
70   AsstProf          A    8        3  Female   78500
17   AsstProf          B    4        0    Male   92000
14       Prof          B   25       19    Male  153750
44       Prof          B   23       19  Female  151768
46   AsstProf          B   11        3  Female   74692
51       Prof          B   20       14  Female  127512
2        Prof          A   23       20    Male  110515
47  AssocProf          B   11       11  Female  103613
74  AssocProf          B   19        6  Female  104542
65       Prof          A   36       19  Female  117555
57   AsstProf          A    3        1  Female   72500
38   AsstProf          B    4        3    Male   91000
32       Prof          A   27       19    Male  103275
37       Prof          A   20        8    Male  102000
29       Prof          B   21       20    Male  123683
30  AssocProf          B    9        7    Male  107008
27       Prof          A   45       43    Male  155865
18       Prof          A   19        7    Male  107300
59  AssocProf          B   12       10  Female  103994
10       Prof          B   39       33    Male  128250
76       Prof          A   28       14  Female  109954
35       Prof          B   28       23    Male  126933
3        Prof          A   40       31    Male  131205
63       Prof          A   29       27  Female   91000
53   AsstProf          A    5        3  Female   73500
36       Prof          B   45       45    Male  146856
67       Prof          A   17       11  Female   90450
49       Prof          B   17       18  Female  122960
9        Prof          A   51       51    Male   57800
19       Prof          A   29       27    Male  150500
25       Prof          B   35       31    Male   99418
58       Prof          B   36       26  Female  144651
0        Prof          B   56       49    Male  186960
20   AsstProf          B    4        4    Male   92000
6   AssocProf          A   20       17    Male   81285
52       Prof          A   12        0  Female  105000
22   AsstProf          A    4        2    Male   73000
4        Prof          B   20       18    Male  104800
21       Prof          A   33       30    Male  103106
71  AssocProf          B   12        9  Female   71065
42   AsstProf          B    4        2  Female   80225
7        Prof          A   18       18    Male  126300
61  AssocProf          B   13       10  Female  103750
77       Prof          A   23       15  Female  109646
8        Prof          A   29       19    Male   94350
23   AsstProf          A    2        0    Male   85000

In [41]:
df_50.salary.mean() # mean of salary of 50  samples 

np.float64(108176.6)

In [47]:
new_df=df_50.select_dtypes(include=[np.number])

In [49]:
new_df.mean()

phd            21.58
service        16.46
salary     108176.60
dtype: float64

In [46]:
new_df.salary.mean()

np.float64(108023.78205128205)

In [50]:
new_df.shape

(50, 3)

In [51]:
df_50.shape

(50, 6)

In [52]:
df.salary

0     186960
1      93000
2     110515
3     131205
4     104800
       ...  
73    105450
74    104542
75    124312
76    109954
77    109646
Name: salary, Length: 78, dtype: int64

In [56]:
new_df.std()

phd           13.840329
service       13.427371
salary     27744.172151
dtype: float64

In [57]:
df.phd.var()

np.float64(156.21062271062272)

In [58]:
df_50.phd.var()

np.float64(191.55469387755102)

In [59]:
new_df.phd.var()

np.float64(191.55469387755102)

to calculate to population varace in python we have to go with numpy 

In [61]:
import numpy as np

In [62]:
np.var(df.phd)

np.float64(154.20792241946089)

SELECTING A COLOUMN IN A DATA SET 


- method 1 : subset the data frame using coloumn name : df['rank ]

- metjod 2 : using coloumn name as an attribute : df.rank

EXAMPLES 

- calulate the basic statistics for phb coloumn :df.phd.describe ()
- find how many values in the phd coloums (use count method ): df.phb.count()
- calculate the average of phb : df.phb.mean()

In [63]:
# DATAFRAME GROUPBY METHOD 

In [64]:
#GROUPBY 


- splecting data into groups based on sertain criteria 
- calculate stats or apple function to each group

In [71]:
df.groupby(['sex'])[["salary","service"]].mean()

salary    service
sex                             
Female  101002.410256  11.564103
Male    115045.153846  18.538462

In [73]:
df.groupby9(['rank',sort=False])[["salary","service"]].mean()

SyntaxError: invalid syntax. Maybe you meant '==' or ':=' instead of '='? (1061131450.py, line 1)